common/s3util.h

```diff
@@ -64,6 +66,61 @@ class S3UtilResponse {
   string error_;
 };
 
+/**
+ * A writable file which uses direct I/O under the hood.
+ */
+class DirectIOFile {
+ public:
+  DirectIOFile(const string& file_path);
+  ~DirectIOFile();
+
+  // no copy or move
+  DirectIOFile(const DirectIOFile&) = delete;
+  DirectIOFile& operator=(const DirectIOFile&) = delete;
+
+  std::streamsize write(const char* s, std::streamsize n);
+
+ private:
+  // file descriptor
+  int fd_;
+  uint32_t file_size_;
+  // page size aligned buffer
+  void* buffer_;
+  // buffer offset
+  uint32_t offset_;
+};
+
+/**
+ * A wrapper class of DirectIOFile which can be used together with
+ * boost::iostreams::stream to implement i{o}stream.
+ */
+class DirectIOFileSink {
+ public:
+  using char_type = char;
+  using category = boost::iostreams::bidirectional_device_tag;
+
+  DirectIOFileSink(const string& file_path)
+      : file_(std::make_shared<DirectIOFile>(file_path)) {
+  }
+
+  std::streamsize write(const char* s, std::streamsize n) {
+    return file_->write(s, n);
+  }
+
+  std::streamsize read(char* s, std::streamsize n) {
```

Comment:
The aws sdk requires a iostream. However, if we don't implement read(), `boost::iostreams::stream<DirectIOFileSink>(local_path)` will return a ostream instead of iostream.